mlrun/run.py

```diff
@@ -1218,3 +1226,609 @@ def wait_for_runs_completion(runs: list, sleep=3, timeout=0, silent=False):
+                                     kwargs.
+        """
+        for obj, log_tuple in zip(outputs, logging_instructions):
+            # Check if needed to log (not None):
+            if log_tuple is None:
+                continue
+            # Log:
+            self._log_output(obj=obj, artifact_type=log_tuple[0], key=log_tuple[1])
+
+    def set_labels(self, labels: Dict[str, str]):
+        """
+        Set the given labels with the stored context.
+
+        :param labels: The labels to set.
+        """
+        for key, value in labels.items():
+            self._context.set_label(key=key, value=value)
+
+    def _parse_input(self, data_item: DataItem, expected_type: Type) -> Any:
+        """
+        Parse the given data frame to the expected type. By default, it will be parsed to an object (will be treated as
+        a pickle).
+
+        :param data_item:     The data item to parse.
+        :param expected_type: THe expected type to parse to.
+
+        :returns: The parsed data item.
+        """
+        return self.PARSER_MAP.get(str(expected_type), self.PARSER_MAP[str(object)])(
+            data_item=data_item
+        )
+
+    def _log_output(
+        self, obj, artifact_type: Union[ArtifactType, str], key: Union[str, dict]
+    ):
+        """
+        Log the given object to MLRun as the given artifact type with the provided key. The key can be part of a
+        logging keyword arguments to pas to the relevant context logging function.
+
+        :param obj:           The object to log.
+        :param artifact_type: The artifact type to log the object as.
+        :param key:           The key (name) of the artifact or a logging kwargs to use when logging the artifact.
+
+        :raises MLRunInvalidArgumentError: If a key was not provided.
+        """
+        # Get the artifact type:
+        artifact_type = ArtifactType(artifact_type)
+
+        # Check if the key is in fact logging kwargs:
+        logging_kwargs = {}
+        if isinstance(key, dict):
+            logging_kwargs = key
+            # Get the key value from the kwargs:
+            key = logging_kwargs.pop("key" if "key" in logging_kwargs else "item", None)
+            if key is None:
+                raise mlrun.errors.MLRunInvalidArgumentError(
+                    "When passing logging keyword arguments, either 'key' or 'item' (according to the context "
+                    "method) must be added to the dictionary."
+                )
+
+        # Use the logging map to log the object:
+        self.LOGGER_MAP[artifact_type.value](
+            context=self._context,
+            obj=obj,
+            key=key,
+            logging_kwargs=logging_kwargs,
+        )
+
+
+def function_decorator(
+    set_labels: Dict[str, str] = None,
+    log_outputs: List[
+        Union[Tuple[Union[ArtifactType, str], Union[str, Dict[str, Any]]], None]
+    ] = None,
```

Comment:
🤯

Reply:
I explained it in the doc string, in my opinion it is better to have type hints than nothing, even if the type hint is long.